## VR Norm

In [1]:
#Data saved in csv/excel
import glob, os
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [2]:
reg = pd.read_csv('C:/Users/tsb7592/Downloads/registration_data_distinct.csv',index_col=0)

In [3]:
reg.shape

(3260, 13)

In [4]:
df = pd.read_csv('C:/Users/tsb7592/Downloads/VR_norm.csv', index_col=0)

# OR Norm & Original Norm

In [5]:
df =df[~df['ItemID'].str.contains('Practice|Title|Intro|Transition')]

In [6]:
df.shape

(49664, 11)

In [7]:
#Importing the FormWords
xls =  pd.read_excel('C:/Users/tsb7592/Downloads/OR Forms for Norming 07.08.2020_WithWords.xlsx', sheet_name = "Formatted_Forms")

In [8]:
df.Key.unique()

array(['InstrumentOrder', 'InstrumentSection', 'ItemOrder', 'Locale',
       'Response', 'Score', 'DataType', 'Position', 'ResponseTime',
       'AppVersion', 'iPadVersion', 'FirmwareVersion'], dtype=object)

In [9]:
#Select the 3 needed key values
df1 =df[(df.Key == 'Response')]
df1['Value'] =pd.to_numeric(df1['Value'])

In [10]:
df2 =df[(df.Key == 'ResponseTime')]
df2['Value'] =pd.to_numeric(df2['Value'])

In [11]:
df3 =df[(df.Key == 'Score')]
df3['Value'] =pd.to_numeric(df3['Value'])

In [12]:
df2.Key.value_counts()

ResponseTime    4138
Name: Key, dtype: int64

In [13]:
df1 = df1[['PIN','RegistrationID','ItemID','Key','Value']].drop_duplicates()
df2 = df2[['PIN','RegistrationID','ItemID','Key','Value']].drop_duplicates()
df3 = df3[['PIN','RegistrationID','ItemID','Key','Value']].drop_duplicates()

In [14]:
df1.shape

(4139, 5)

In [93]:
correct = df3[df3.Value == 1]

In [102]:
wrong= df3[df3.Value == 0]

In [103]:
correctppl =pd.merge(correct, df2, on=['ItemID','PIN'] )
wrongppl =pd.merge(wrong, df2, on=['ItemID','PIN'] )

In [104]:
correctppl.columns =['PIN', 'RegistrationID', 'ItemID', 'Key_x', 'Value_x',
       'RegistrationID_y', 'Key_y', 'Time']
wrongppl.columns =['PIN', 'RegistrationID', 'ItemID', 'Key_x', 'Value_x',
       'RegistrationID_y', 'Key_y', 'Time']


In [96]:
cp = correctppl.pivot_table(index = ['PIN', 'RegistrationID'], columns='ItemID',values='Time').reset_index()

In [105]:
wp = wrongppl.pivot_table(index = ['PIN', 'RegistrationID'], columns='ItemID',values='Time').reset_index()
wp

ItemID       PIN                        RegistrationID   VA_RA022  VR_MR001  \
0            MAP  A2120EFC-4C20-4014-B304-290EB08DDF8B        NaN       NaN   
1       MAP03001  A73D460E-532B-4BF6-B7AE-4EAC5DF50EBC        NaN       NaN   
2       MAP03002  B4734787-856D-49DA-B952-18885F6A3EA0  21.952430       NaN   
3       MAP03003  134B76E9-6FB1-41BE-BDF3-B31B49C4C790   0.585175       NaN   
4       MAP03005  AD6B2CCD-76A6-4A7D-924A-13FE7B4D6988  19.721380       NaN   
..           ...                                   ...        ...       ...   
181     MAP70121  143E2594-996C-439A-9242-02C0ED68674A        NaN       NaN   
182     MAP70122  FF73BBBA-546D-4303-B9CD-0F1526FB22E8        NaN       NaN   
183     MAP80002  23651D9D-2A8D-49FD-9CC4-E9FF01324FE5        NaN  3.016118   
184     MAP80003  38F883B0-0881-4DD6-8736-AD69F9D7B400        NaN       NaN   
185     MAP85003  B8B9ACC7-F032-44F9-A27D-60E0319EE06C        NaN       NaN   

ItemID  VR_MR002  VR_MR003   VR_MR004  VR_MR005  VR_MR006  VR_MR007  ...  \
0            NaN       NaN        NaN       NaN       NaN       NaN  ...   
1            NaN       NaN        NaN       NaN       NaN       NaN  ...   
2            NaN       NaN        NaN       NaN       NaN       NaN  ...   
3            NaN       NaN        NaN       NaN       NaN       NaN  ...   
4            NaN       NaN        NaN       NaN       NaN       NaN  ...   
..           ...       ...        ...       ...       ...       ...  ...   
181          NaN       NaN  18.621760       NaN       NaN  71.99805  ...   
182          NaN       NaN        NaN       NaN       NaN       NaN  ...   
183          NaN       NaN        NaN       NaN       NaN       NaN  ...   
184          NaN       NaN   6.150751       NaN       NaN       NaN  ...   
185          NaN       NaN        NaN       NaN       NaN       NaN  ...   

ItemID  VR_SV003  VR_SV004R1  VR_SV005R1  VR_SV006R1  VR_SV007R1  VR_SV008R1  \
0            NaN         NaN         NaN         NaN         NaN         NaN   
1            NaN         NaN         NaN         NaN         NaN         NaN   
2            NaN         NaN         NaN         NaN         NaN         NaN   
3            NaN         NaN         NaN         NaN         NaN         NaN   
4            NaN         NaN         NaN         NaN         NaN         NaN   
..           ...         ...         ...         ...         ...         ...   
181          NaN         NaN         NaN         NaN         NaN         NaN   
182          NaN         NaN         NaN         NaN         NaN         NaN   
183          NaN         NaN         NaN         NaN         NaN         NaN   
184          NaN    5.624436         NaN         NaN         NaN         NaN   
185          NaN         NaN         NaN         NaN     15.6306         NaN   

ItemID  VR_SV009  VR_SV010  VR_SV011  VR_SV012  
0            NaN       NaN       NaN       NaN  
1            NaN       NaN       NaN       NaN  
2            NaN       NaN       NaN       NaN  
3            NaN       NaN       NaN       NaN  
4            NaN       NaN       NaN       NaN  
..           ...       ...       ...       ...  
181          NaN       NaN       NaN       NaN  
182          NaN       NaN       NaN       NaN  
183          NaN       NaN       NaN       NaN  
184     6.790805       NaN       NaN       NaN  
185          NaN       NaN       NaN       NaN  

[186 rows x 192 columns]

In [15]:
vr1=pd.merge(df1, reg[['PIN','Age']], on=['PIN'], how='left')
vr2=pd.merge(df2, reg[['PIN','Age']], on=['PIN'], how='left')
vr3=pd.merge(df3, reg[['PIN','Age']], on=['PIN'], how='left')

In [116]:
vr1.ItemID.nunique()

194

In [17]:
#long to wide
VR_Response =vr1.pivot_table(index = ['PIN', 'RegistrationID','Age'], columns='ItemID',values='Value').reset_index()

In [18]:
VR_ResponseTime =vr2.pivot_table(index = ['PIN','RegistrationID', 'Age'], columns='ItemID',values='Value').reset_index()

In [19]:
VR_Score =vr3.pivot_table(index = ['PIN', 'RegistrationID','Age'], columns='ItemID',values='Value').reset_index()

In [20]:
VR_ResponseTime.to_csv('C:/Users/tsb7592/Downloads/VR_ResponseTime.csv',index=False)

In [21]:
VR_Response.to_csv('C:/Users/tsb7592/Downloads/VR_Response.csv',index=False)

In [163]:
VR_Score.to_csv('C:/Users/tsb7592/Downloads/VR_Score.csv',index=False)

# Analysis
## Reaction time: 


In [162]:
VR_Score.VR_MR001

0      NaN
1      NaN
2      NaN
3      NaN
4      NaN
      ... 
183    1.0
184    NaN
185    0.0
186    NaN
187    NaN
Name: VR_MR001, Length: 188, dtype: float64

In [169]:
df[(df.ItemID == 'VR_MR001') & (df.Key=='Score')].shape

(26, 11)

In [43]:
VR_Score.describe()

ItemID         Age   VA_RA022   VR_MR001   VR_MR002   VR_MR003   VR_MR004  \
count   188.000000  13.000000  26.000000  16.000000  17.000000  18.000000   
mean     22.925532   0.153846   0.500000   0.187500   0.294118   0.111111   
std      21.970198   0.375534   0.509902   0.403113   0.469668   0.323381   
min       3.000000   0.000000   0.000000   0.000000   0.000000   0.000000   
25%       8.000000   0.000000   0.000000   0.000000   0.000000   0.000000   
50%      14.000000   0.000000   0.500000   0.000000   0.000000   0.000000   
75%      31.000000   0.000000   1.000000   0.000000   1.000000   0.000000   
max      86.000000   1.000000   1.000000   1.000000   1.000000   1.000000   

ItemID   VR_MR005   VR_MR006   VR_MR007  VR_MR008  ...   VR_SV003  VR_SV004R1  \
count   22.000000  24.000000  15.000000      16.0  ...  22.000000   30.000000   
mean     0.636364   0.583333   0.266667       0.0  ...   0.590909    0.700000   
std      0.492366   0.503610   0.457738       0.0  ...   0.503236    0.466092   
min      0.000000   0.000000   0.000000       0.0  ...   0.000000    0.000000   
25%      0.000000   0.000000   0.000000       0.0  ...   0.000000    0.000000   
50%      1.000000   1.000000   0.000000       0.0  ...   1.000000    1.000000   
75%      1.000000   1.000000   0.500000       0.0  ...   1.000000    1.000000   
max      1.000000   1.000000   1.000000       0.0  ...   1.000000    1.000000   

ItemID  VR_SV005R1  VR_SV006R1  VR_SV007R1  VR_SV008R1   VR_SV009   VR_SV010  \
count    16.000000   27.000000   19.000000   24.000000  18.000000  23.000000   
mean      0.562500    0.518519    0.315789    0.041667   0.555556   0.521739   
std       0.512348    0.509175    0.477567    0.204124   0.511310   0.510754   
min       0.000000    0.000000    0.000000    0.000000   0.000000   0.000000   
25%       0.000000    0.000000    0.000000    0.000000   0.000000   0.000000   
50%       1.000000    1.000000    0.000000    0.000000   1.000000   1.000000   
75%       1.000000    1.000000    1.000000    0.000000   1.000000   1.000000   
max       1.000000    1.000000    1.000000    1.000000   1.000000   1.000000   

ItemID   VR_SV011   VR_SV012  
count   22.000000  23.000000  
mean     0.318182   0.652174  
std      0.476731   0.486985  
min      0.000000   0.000000  
25%      0.000000   0.000000  
50%      0.000000   1.000000  
75%      1.000000   1.000000  
max      1.000000   1.000000  

[8 rows x 195 columns]

In [23]:
##For each item, how long did it take people on average to complete the item? 
count =VR_ResponseTime.describe().round(2).iloc[0]
meantime =VR_ResponseTime.describe().round(2).iloc[1]

In [24]:
rv = pd.concat([count,meantime], axis=1)
rv =rv.iloc[1:]

In [25]:
rv.columns =['N', 'MeanTime_Overall']

In [26]:
rv =rv.reset_index()

In [27]:
timecol = VR_ResponseTime.columns[3:].to_list()

In [28]:
timecol

['VA_RA022',
 'VR_MR001',
 'VR_MR002',
 'VR_MR003',
 'VR_MR004',
 'VR_MR005',
 'VR_MR006',
 'VR_MR007',
 'VR_MR008',
 'VR_MR009R1',
 'VR_MR010',
 'VR_MR011R1',
 'VR_MR012',
 'VR_MR013R1',
 'VR_MR014R1',
 'VR_MR015R1',
 'VR_MR016',
 'VR_MR017',
 'VR_MR018',
 'VR_MR019R1',
 'VR_MR020',
 'VR_MR021',
 'VR_MR022',
 'VR_MR023',
 'VR_MR024',
 'VR_MR025R1',
 'VR_MR026R1',
 'VR_MR027R1',
 'VR_MR028',
 'VR_MR029',
 'VR_MR030R1',
 'VR_MR031R1',
 'VR_MR032',
 'VR_MR033R1',
 'VR_MR034R1',
 'VR_MR035',
 'VR_MR036',
 'VR_MR037',
 'VR_MR038',
 'VR_MR039R1',
 'VR_PC001R1',
 'VR_PC002',
 'VR_PC004R1',
 'VR_PC005R1',
 'VR_PC006R1',
 'VR_PC007',
 'VR_PC008',
 'VR_PC009',
 'VR_PC011',
 'VR_PC012R1',
 'VR_PC013',
 'VR_PC014R1',
 'VR_PC015',
 'VR_PC016',
 'VR_PC017',
 'VR_PC018R1',
 'VR_PC019R1',
 'VR_PC020',
 'VR_PC021R1',
 'VR_PC022R1',
 'VR_PC023',
 'VR_PC024',
 'VR_PC025',
 'VR_PC026',
 'VR_PC027',
 'VR_RA001',
 'VR_RA002R1',
 'VR_RA004',
 'VR_RA011R1',
 'VR_RA013',
 'VR_RA014R1',
 'VR_RA015R1',
 'VR_RA0

In [29]:
adc ={}
for col in timecol:
    adc[col]=round(VR_ResponseTime[VR_ResponseTime[col].notnull()]['Age'].mean(),2)

In [30]:
meanage = pd.DataFrame.from_dict(adc,orient='index')

In [31]:
meanage=meanage.reset_index()
meanage.columns=['ItemID','mean_age']

In [36]:
meanage

ItemID  mean_age
0      VA_RA022      3.46
1      VR_MR001     29.31
2      VR_MR002     35.06
3      VR_MR003     30.82
4      VR_MR004     37.39
..          ...       ...
189  VR_SV008R1     39.08
190    VR_SV009     35.61
191    VR_SV010     34.26
192    VR_SV011     27.09
193    VR_SV012     25.35

[194 rows x 2 columns]

In [41]:
data=pd.merge(meanage, rv, on="ItemID")

In [141]:
data.head()

ItemID  mean_age     N  MeanTime_Overall
0  VA_RA022      3.46  13.0             11.13
1  VR_MR001     29.31  26.0              9.09
2  VR_MR002     35.06  16.0             10.08
3  VR_MR003     30.82  17.0             16.91
4  VR_MR004     37.39  18.0             17.98


Within each item, how long did it take people on average to complete the item if they got it right, and if they got it wrong? 

So, the resulting dataset should be approx 195 rows and 9 columns. For the 6th column you may need to join the Response dataset you created earlier to get this column, so maybe do it at the very end. 

In [142]:
VR_Score.describe().round(2)

ItemID     Age  VA_RA022  VR_MR001  VR_MR002  VR_MR003  VR_MR004  VR_MR005  \
count   188.00     13.00     26.00     16.00     17.00     18.00     22.00   
mean     22.93      0.15      0.50      0.19      0.29      0.11      0.64   
std      21.97      0.38      0.51      0.40      0.47      0.32      0.49   
min       3.00      0.00      0.00      0.00      0.00      0.00      0.00   
25%       8.00      0.00      0.00      0.00      0.00      0.00      0.00   
50%      14.00      0.00      0.50      0.00      0.00      0.00      1.00   
75%      31.00      0.00      1.00      0.00      1.00      0.00      1.00   
max      86.00      1.00      1.00      1.00      1.00      1.00      1.00   

ItemID  VR_MR006  VR_MR007  VR_MR008  ...  VR_SV003  VR_SV004R1  VR_SV005R1  \
count      24.00     15.00      16.0  ...     22.00       30.00       16.00   
mean        0.58      0.27       0.0  ...      0.59        0.70        0.56   
std         0.50      0.46       0.0  ...      0.50        0.47        0.51   
min         0.00      0.00       0.0  ...      0.00        0.00        0.00   
25%         0.00      0.00       0.0  ...      0.00        0.00        0.00   
50%         1.00      0.00       0.0  ...      1.00        1.00        1.00   
75%         1.00      0.50       0.0  ...      1.00        1.00        1.00   
max         1.00      1.00       0.0  ...      1.00        1.00        1.00   

ItemID  VR_SV006R1  VR_SV007R1  VR_SV008R1  VR_SV009  VR_SV010  VR_SV011  \
count        27.00       19.00       24.00     18.00     23.00     22.00   
mean          0.52        0.32        0.04      0.56      0.52      0.32   
std           0.51        0.48        0.20      0.51      0.51      0.48   
min           0.00        0.00        0.00      0.00      0.00      0.00   
25%           0.00        0.00        0.00      0.00      0.00      0.00   
50%           1.00        0.00        0.00      1.00      1.00      0.00   
75%           1.00        1.00        0.00      1.00      1.00      1.00   
max           1.00        1.00        1.00      1.00      1.00      1.00   

ItemID  VR_SV012  
count      23.00  
mean        0.65  
std         0.49  
min         0.00  
25%         0.00  
50%         1.00  
75%         1.00  
max         1.00  

[8 rows x 195 columns]

In [143]:
##proportion of people who got it right (sum of those values == 1 divided by N)
proportionRight =VR_Score.describe().round(2).iloc[1]

In [144]:
proportionRight =proportionRight.iloc[1:]

In [145]:
cp

ItemID       PIN                        RegistrationID  VA_RA022  VR_MR001  \
0            MAP  A2120EFC-4C20-4014-B304-290EB08DDF8B       NaN       NaN   
1       MAP03001  A73D460E-532B-4BF6-B7AE-4EAC5DF50EBC       NaN       NaN   
2       MAP03002  B4734787-856D-49DA-B952-18885F6A3EA0       NaN       NaN   
3       MAP03003  134B76E9-6FB1-41BE-BDF3-B31B49C4C790       NaN       NaN   
4       MAP03005  AD6B2CCD-76A6-4A7D-924A-13FE7B4D6988       NaN       NaN   
..           ...                                   ...       ...       ...   
181     MAP70121  143E2594-996C-439A-9242-02C0ED68674A       NaN  24.31291   
182     MAP70122  FF73BBBA-546D-4303-B9CD-0F1526FB22E8       NaN       NaN   
183     MAP80002  23651D9D-2A8D-49FD-9CC4-E9FF01324FE5       NaN       NaN   
184     MAP80003  38F883B0-0881-4DD6-8736-AD69F9D7B400       NaN       NaN   
185     MAP85003  B8B9ACC7-F032-44F9-A27D-60E0319EE06C       NaN       NaN   

ItemID  VR_MR002  VR_MR003  VR_MR004  VR_MR005  VR_MR006  VR_MR007  ...  \
0            NaN       NaN       NaN  2.692112       NaN       NaN  ...   
1            NaN       NaN       NaN       NaN       NaN       NaN  ...   
2            NaN       NaN       NaN       NaN       NaN       NaN  ...   
3            NaN       NaN       NaN       NaN       NaN       NaN  ...   
4            NaN       NaN       NaN       NaN       NaN       NaN  ...   
..           ...       ...       ...       ...       ...       ...  ...   
181          NaN       NaN       NaN       NaN       NaN       NaN  ...   
182          NaN       NaN       NaN       NaN       NaN       NaN  ...   
183          NaN       NaN       NaN       NaN       NaN       NaN  ...   
184          NaN       NaN       NaN       NaN       NaN       NaN  ...   
185          NaN       NaN       NaN       NaN       NaN       NaN  ...   

ItemID  VR_SV003  VR_SV004R1  VR_SV005R1  VR_SV006R1  VR_SV007R1  VR_SV008R1  \
0            NaN         NaN         NaN         NaN         NaN         NaN   
1            NaN         NaN         NaN         NaN         NaN         NaN   
2            NaN         NaN         NaN         NaN         NaN         NaN   
3            NaN         NaN         NaN         NaN         NaN         NaN   
4            NaN         NaN         NaN         NaN         NaN         NaN   
..           ...         ...         ...         ...         ...         ...   
181          NaN         NaN         NaN         NaN         NaN         NaN   
182          NaN         NaN         NaN         NaN         NaN         NaN   
183          NaN         NaN    4.999264         NaN         NaN         NaN   
184     4.105292         NaN         NaN         NaN         NaN         NaN   
185          NaN    7.071551         NaN    8.808719         NaN         NaN   

ItemID  VR_SV009  VR_SV010  VR_SV011  VR_SV012  
0            NaN       NaN  1.792682  4.783233  
1            NaN       NaN       NaN       NaN  
2            NaN       NaN       NaN       NaN  
3            NaN       NaN       NaN       NaN  
4            NaN       NaN       NaN       NaN  
..           ...       ...       ...       ...  
181          NaN       NaN       NaN       NaN  
182     17.97998       NaN       NaN       NaN  
183          NaN       NaN       NaN       NaN  
184          NaN       NaN       NaN       NaN  
185          NaN  8.240641       NaN       NaN  

[186 rows x 195 columns]

In [146]:
#within those people who got it right, what was their mean 
meantime_c = cp.describe().round(2).iloc[1]
#within those people who got it right, what was their SD
sdtime_c = cp.describe().round(2).iloc[2]

In [147]:
correctcol = pd.concat([proportionRight,meantime_c, sdtime_c], axis=1)

In [148]:
correctcol.columns = ['ProportionCorrect', 'MeanTime_Correct', 'SDTime_Correct']

In [149]:
datacorrect =pd.merge(data,correctcol, how='left',on="ItemID")

In [150]:
cp[cp['VR_SV008R1'].notnull()]

ItemID       PIN                        RegistrationID  VA_RA022  VR_MR001  \
169     MAP60015  FF15E5E2-E345-4F09-9C59-AFE005F8D4BC       NaN       NaN   

ItemID  VR_MR002  VR_MR003  VR_MR004  VR_MR005  VR_MR006  VR_MR007  ...  \
169          NaN       NaN       NaN       NaN       NaN       NaN  ...   

ItemID  VR_SV003  VR_SV004R1  VR_SV005R1  VR_SV006R1  VR_SV007R1  VR_SV008R1  \
169          NaN         NaN    10.48826         NaN         NaN    45.18937   

ItemID  VR_SV009  VR_SV010  VR_SV011  VR_SV012  
169          NaN       NaN       NaN       NaN  

[1 rows x 195 columns]

In [151]:
#within those people who got it wrong, what was their mean 
meantime_w = wp.describe().round(2).iloc[1]
#within those people who got it wrong, what was their SD
sdtime_w = wp.describe().round(2).iloc[2]

In [152]:
wcol = pd.concat([meantime_w, sdtime_w], axis=1)

In [153]:
wcol.columns = [ 'MeanTime_Incorrect', 'SDTime_Incorrect']

In [154]:
final =pd.merge(datacorrect, wcol, how='left',on="ItemID")

In [155]:
final.shape

(194, 9)

In [156]:
common = final.merge(vr1['ItemID'],on=['ItemID'], how="right")

In [157]:
common.drop_duplicates()

ItemID  mean_age     N  MeanTime_Overall  ProportionCorrect  \
0     VR_RA058R1     23.86  35.0              8.85               0.54   
1    VR_SS3004R1      7.46  13.0              4.28               0.31   
2       VR_PC002      7.28  18.0              5.13               0.67   
3    VR_SS3005R1     21.13  38.0              9.34               0.79   
4     VR_RA062R1     23.20  35.0              9.65               0.49   
..           ...       ...   ...               ...                ...   
684     VR_RA001      3.44   9.0              5.42               0.44   
688   VR_RA044R1      3.36  11.0              4.72               0.55   
710   VR_RA011R1      3.50   8.0              4.80               0.12   
805   VR_MR025R1     31.68  19.0             13.48               0.42   
815  VR_SS2111R1     28.94  17.0              5.96               0.35   

     MeanTime_Correct  SDTime_Correct  MeanTime_Incorrect  SDTime_Incorrect  
0                7.87            3.53               10.02              7.59  
1                4.38            1.87                4.23              1.41  
2                5.45            1.99                4.49              2.13  
3                8.28            4.04               13.32              9.87  
4                9.84            5.32                9.46              4.76  
..                ...             ...                 ...               ...  
684              5.81            2.76                5.11              5.77  
688              5.97            4.72                3.23              2.34  
710              2.25             NaN                5.17              3.12  
805             16.68            6.93               11.15              8.90  
815              8.38            4.41                4.64              2.17  

[194 rows x 9 columns]

In [158]:
final[final.ItemID == 'VR_PC013']

ItemID  mean_age     N  MeanTime_Overall  ProportionCorrect  \
50  VR_PC013       8.0  11.0              6.14                1.0   

    MeanTime_Correct  SDTime_Correct  MeanTime_Incorrect  SDTime_Incorrect  
50              6.14            2.37                 NaN               NaN

In [160]:
final.to_csv('C:/Users/tsb7592/Downloads/Reaction_time.csv', index=False)

In [161]:
final.shape

(194, 9)